libraries/chain/database.cpp

```diff
@@ -2005,6 +2014,94 @@ void database::apply_operation(const operation& op)
    notify_post_apply_operation( note );
 }
 
+void database::process_tx_bandwidth(const signed_transaction& trx) {
+   account_bandwidth_object trx_bandwidth_data;
+
+   trx_bandwidth_data.total_bandwidth = fc::raw::pack_size(trx);
+   trx_bandwidth_data.total_ops_count = trx.operations.size();
+
+   bool fee_free_op_present = false;
+   bool paid_op_present     = false;
+   for (const operation& op : trx.operations) {
+      if (is_fee_free_operation(op) == true) {
+         // Count standalone fee-free operations bandwidth
+         trx_bandwidth_data.act_fee_free_bandwidth += fc::raw::pack_size(op);
+         trx_bandwidth_data.act_fee_free_ops_count++;
+         fee_free_op_present = true;
+      }
+      else {
+         paid_op_present = true;
+      }
+   }
+
+   // If there are only fee-free operations, count as consumed fee-free bandwidth also transaction meta info,
+   // if there are also paid operations, count as consumed fee-free bandwidth ONLY standalone operations(see above)
+   if (paid_op_present == false && fee_free_op_present == true) {
+      trx_bandwidth_data.act_fee_free_bandwidth = trx_bandwidth_data.total_bandwidth;
+   }
+
+
+   flat_set< account_name_type > required; vector<authority> other;
+   trx.get_required_authorities( required, required, other );
+
+   // Updates all involved accounts bandwidth
+   for( const auto& auth : required )
+   {
+      const auto& account = this->get_account( auth );
+      update_account_bandwidth( account.name, trx_bandwidth_data, fee_free_op_present);
+   }
+}
+
+void database::update_account_bandwidth(const account_name_type& account, const account_bandwidth_object& trx_bandwidth_data, bool fee_free_op_present) {
+   // Adjusts account bandwidth statistics
+   const account_bandwidth_object* acc_bandwidth = this->find<account_bandwidth_object, by_account>(account);
+   uint32_t act_head_block = this->head_block_num();
+
+   if (acc_bandwidth == nullptr) {
```

Comment:
Can this whole if/else block be moved into `if (fee_free_op_present == true) ` block to do lazy evaluation/avoid processing and updating when there is no free operation involved?

Reply:
it can. But i count also total bandwdith, which includes also paid operations... Fought it might be interesting to see such information...